src/components/MalloyQueryFocusProvider.tsx

```diff
@@ -0,0 +1,141 @@
+import {
+  ASTArrowQueryDefinition,
+  ASTArrowViewDefinition,
+  ASTNestViewOperation,
+  ASTQuery,
+  ASTRefinementViewDefinition,
+  ASTSegmentViewDefinition,
+  ASTView,
+  ASTViewDefinition,
+} from '@malloydata/malloy-query-builder';
+
+interface MalloyQueryFocus {
+  focusMainView: () => void;
+  isMainViewFocused: boolean;
+  focusNestView: (path: string[]) => void;
+  isNestViewFocused: (path: string[]) => boolean;
+  focusedNestView: ASTView | null;
+}
+
+const MalloyQueryFocusContext = React.createContext<MalloyQueryFocus>({
+  focusMainView: () => {},
+  isMainViewFocused: true,
+  focusNestView: () => {},
+  isNestViewFocused: () => false,
+  focusedNestView: null,
+});
+
+interface MalloyQueryFocusProviderProps {
+  rootQuery: ASTQuery | undefined;
+  children: ReactNode | ReactNode[];
+}
+
+export function MalloyQueryFocusProvider({
+  rootQuery,
+  children,
+}: MalloyQueryFocusProviderProps) {
+  const [focusedNestViewPath, setFocusedNestViewPath] = React.useState<
+    string[]
+  >([]);
+
+  const focusedNestView = React.useMemo<ASTView | null>(() => {
+    if (focusedNestViewPath.length === 0) {
+      return null;
+    }
+
+    if (rootQuery) {
+      const queryDef = rootQuery.definition;
+      if (queryDef instanceof ASTArrowQueryDefinition) {
+        return findNestView(queryDef.view, [...focusedNestViewPath].reverse());
+      }
+    }
+
+    return null;
+  }, [rootQuery, focusedNestViewPath]);
+
+  const focusMainView = React.useCallback(() => {
+    setFocusedNestViewPath([]);
+  }, []);
+
+  const focusNestView = React.useCallback((path: string[]) => {
+    setFocusedNestViewPath([...path]);
+  }, []);
+
+  const isNestViewFocused = React.useCallback(
+    (path: string[]) => {
+      return JSON.stringify(path) === JSON.stringify(focusedNestViewPath);
+    },
+    [focusedNestViewPath]
+  );
+
+  const isMainViewFocused = focusedNestViewPath.length === 0;
+
+  return (
+    <MalloyQueryFocusContext.Provider
+      value={{
+        focusMainView,
+        isMainViewFocused,
+        focusNestView,
+        isNestViewFocused,
+        focusedNestView,
+      }}
+    >
+      {children}
+    </MalloyQueryFocusContext.Provider>
+  );
+}
+
+export function useQueryFocus() {
+  return React.useContext(MalloyQueryFocusContext);
+}
+
+const findNestView = (
+  currentView: ASTViewDefinition,
+  remainingPath: string[]
+): ASTView | null => {
+  if (remainingPath.length === 0) {
+    return null;
+  }
+  if (currentView instanceof ASTArrowViewDefinition) {
+    return findNestView(currentView.view, remainingPath);
+  }
+  if (currentView instanceof ASTSegmentViewDefinition) {
+    const nestViewOperations = currentView.operations.items.filter(
+      operation => operation instanceof ASTNestViewOperation
+    );
+    const targetNestName = remainingPath.pop() as string;
+    const targetNestViewOperations = nestViewOperations.filter(
+      operation => operation.name === targetNestName
+    );
```

Comment:
These two `filter()`s could be rolled into one `find()`. There should never be more than one nest with a given name.